### PR TITLE
feat(mark-as-bad): add comment, show comment

### PR DESCRIPTION
**Pull request by emjburns**
Closes https://github.com/spinnaker/keel/issues/1233.

Adds a table to store 'mark as bad' comments + user + timestamp. 

`POST /applications/{application}/veto`
with body:

```kotlin
data class EnvironmentArtifactVeto(
  val targetEnvironment: String,
  val reference: String,
  val version: String,
  val comment: String?
)
```
Marks an artifact as bad and saves the user who vetoed.

Unmark as bad is `DELETE /{application}/veto/{targetEnvironment}/{reference}/{version}`

(will make the corresponding updates in `gate` shortly)

### CODE REVIEW

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt [hunk 1]
```diff
@@ -121,9 +121,14 @@ class ResourceActuator(
...
-                    artifact = artifact,
-                    version = artifactVersion,
-                    targetEnvironment = environment)
+                    veto = EnvironmentArtifactVeto(
+                      reference = artifact.reference,
+                      version = artifactVersion,
+                      targetEnvironment = environment,
+                      vetoedBy = "Spinnaker",
+                      comment = "Automatically vetoed because deployment of this version failed."
+                    )
```
Comment (emjburns) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt:130]: This creates an automated user + reason for automatic rollbacks. Thoughts on wording here?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/api/ArtifactSummary.kt [hunk 2]
```diff
@@ -58,12 +58,14 @@ data class ArtifactSummaryInEnvironment(
...
-  val pinned: Pinned? = null,
+  val pinned: ActionMetadata? = null,
+  val isVetoed: Boolean = false,
+  val vetoed: ActionMetadata? = null,
   val statefulConstraints: List<StatefulConstraintSummary> = emptyList(),
   val statelessConstraints: List<StatelessConstraintSummary> = emptyList()
 )
 
-data class Pinned(
+data class ActionMetadata(
```
Comment (emjburns) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/api/ArtifactSummary.kt:68]: Reusing across pin + veto

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt [hunk 3]
```diff
@@ -952,15 +974,28 @@ class SqlArtifactRepository(
...
         .orderBy(ENVIRONMENT_ARTIFACT_VERSIONS.DEPLOYED_AT.desc())
         .fetchOne { (version, deployedAt, promotionStatus, replacedBy, replacedAt) ->
-          val pinned: Pinned? = jooq
-            .select(ENVIRONMENT_ARTIFACT_PIN.PINNED_BY, ENVIRONMENT_ARTIFACT_PIN.PINNED_AT, ENVIRONMENT_ARTIFACT_PIN.COMMENT)
-            .from(ENVIRONMENT_ARTIFACT_PIN)
-            .where(ENVIRONMENT_ARTIFACT_PIN.ENVIRONMENT_UID.eq(deliveryConfig.getUidFor(environmentName)))
-            .and(ENVIRONMENT_ARTIFACT_PIN.ARTIFACT_UID.eq(artifact.uid))
-            .and(ENVIRONMENT_ARTIFACT_PIN.ARTIFACT_VERSION.eq(version))
-            .fetchOne { (pinnedBy, pinnedAt, comment) ->
-              Pinned(at = pinnedAt.toInstant(UTC), by = pinnedBy, comment = comment)
```
Comment (emjburns) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt:962]: I'm worried that gathering this data will make the api slower. I'm not going to fix that in this PR, I think we need a broader discussion about how to cache state here or generate api view that get get refreshed.

File: keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/ApplicationController.kt [hunk 4]
```diff
@@ -198,8 +199,10 @@ class ApplicationController(
     @PathVariable("application") application: String,
     @RequestBody veto: EnvironmentArtifactVeto
   ) {
-
-    applicationService.markAsVetoedIn(application, veto, true)
+    val succeeded = applicationService.markAsVetoedIn(user, application, veto, true)
+    if (!succeeded) {
+      throw InvalidVetoException(application, veto.targetEnvironment, veto.reference, veto.version)
+    }
```
Comment (emjburns) [inline keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/ApplicationController.kt:205]: a 'mark as bad' could fail if that version is pinned. in this case, we throw an error with a good user message.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt (hunk 1 above)
Comment (gal-yardeni) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt:130]: Maybe "this version is vetoed automatically as we tried to deploy it multiple times and it failed"?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/api/ArtifactSummary.kt (hunk 2 above)
Comment (gal-yardeni) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/api/ArtifactSummary.kt:68]: Love it!

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/api/EnvironmentArtifactVeto.kt [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package com.netflix.spinnaker.keel.core.api
+
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import java.time.Instant
+
```
Comment (gal-yardeni) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/api/EnvironmentArtifactVeto.kt:5]: I'm confused about the 3 classes here -- can you please add some descriptions?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/exceptions/InvalidVetoException.kt [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package com.netflix.spinnaker.keel.exceptions
+
+import org.omg.CORBA.UserException
+
+class InvalidVetoException(application: String, environment: String, reference: String, version: String) :
+  UserException("Unable to veto artifact version $version with reference $reference in environment " +
+  "$environment and application $application. Is this version pinned?")
```
Comment (gal-yardeni) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/exceptions/InvalidVetoException.kt:7]: Is pinning the only reason to get this exception?

File: keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/ApplicationController.kt (hunk 4 above)
Comment (gal-yardeni) [reply keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/ApplicationController.kt:205]: is it possible to throw that exception within `pplicationService.markAsVetoedIn`? it'll make the controller cleaner.

File: keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/ExceptionHandler.kt [hunk 7]
```diff
@@ -86,9 +87,9 @@ class ExceptionHandler(
     return ApiError(CONFLICT, e)
   }
 
-  @ExceptionHandler(ValidationException::class)
+  @ExceptionHandler(ValidationException::class, InvalidVetoException::class)
   @ResponseStatus(BAD_REQUEST)
-  fun onInvalidDeliveryConfig(e: ValidationException): ApiError {
+  fun onInvalidDeliveryConfig(e: Exception): ApiError {
```
Comment (gal-yardeni) [inline keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/ExceptionHandler.kt:92]: why was this changed?

Comment (emjburns) [reply keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/ExceptionHandler.kt:92]: So that we could do the same thing for multiple exception classes (previously this was only handling the `ValidationException` as a `BAD_REQUEST`

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/api/EnvironmentArtifactVeto.kt (hunk 5 above)
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/api/EnvironmentArtifactVeto.kt:5]: I can for the first one, i'll remove the middle one (i didn't need it but forgot about it), and the third is just a lift from another class to not be in with the pin stuff

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/exceptions/InvalidVetoException.kt (hunk 6 above)
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/exceptions/InvalidVetoException.kt:7]: I think there might be another way (like something fails?) but this is the most probable way, and as far as I could tell the only way the 'veto' function would return false (false == veto didn't work). That's why I kept this like asking a question

File: keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/ApplicationController.kt (hunk 4 above)
Comment (emjburns) [reply keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/ApplicationController.kt:205]: sure

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/api/ArtifactSummary.kt [hunk 8]
```diff
@@ -58,12 +58,14 @@ data class ArtifactSummaryInEnvironment(
   val replacedAt: Instant? = null,
   val replacedBy: String? = null,
   val isPinned: Boolean = false,
-  val pinned: Pinned? = null,
+  val pinned: ActionMetadata? = null,
+  val isVetoed: Boolean = false,
```
Comment (lorin) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/api/ArtifactSummary.kt:62]: Is it ever possible for `isVetoed` to be true and `vetoed` to be null? If not, I would just change `isVetoed` to a property that checks if `vetoed` is null. (Same with `pinned`)

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt (hunk 1 above)
Comment (lorin) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt:130]: I like @gal-yardeni's idea of communicating that there were multiple attempts.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt [hunk 9]
```diff
@@ -102,18 +103,17 @@ class ApplicationService(
...
-      artifact = artifact,
-      version = veto.version,
-      targetEnvironment = veto.targetEnvironment,
+      veto = veto.copy(vetoedBy = user),
       force = force
     )
+    if (!succeeded) {
+      throw InvalidVetoException(application, veto.targetEnvironment, veto.reference, veto.version)
+    }
+    // TODO: publish ArtifactVetoedEvent
```
Comment (lorin) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt:116]: Is this `TODO` something that should be done for this PR, or can it come later?

Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt:116]: no, it's for later. the best kind of todo :D 

We need to surface events other than resource events, and we don't do that now. So I don't want to add any events

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt [hunk 10]
```diff
@@ -574,33 +576,46 @@ class SqlArtifactRepository(
...
+                ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION.eq(veto.version)
               )
               .execute()
 
+            txn.insertInto(ENVIRONMENT_ARTIFACT_VETO)
+              .set(ENVIRONMENT_ARTIFACT_VETO.ENVIRONMENT_UID, envUid)
+              .set(ENVIRONMENT_ARTIFACT_VETO.ARTIFACT_UID, artUid)
+              .set(ENVIRONMENT_ARTIFACT_VETO.ARTIFACT_VERSION, veto.version)
+              .set(ENVIRONMENT_ARTIFACT_VETO.VETOED_AT, clock.timestamp())
+              .set(ENVIRONMENT_ARTIFACT_VETO.VETOED_BY, veto.vetoedBy)
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt:611]: I didn't want to have a separate class just to capture that this is not null after it's submitted. I think the risk of this throwing an exception is pretty low. thanks for pointing it out though!

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/api/ArtifactSummary.kt (hunk 8 above)
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/api/ArtifactSummary.kt:62]: That's actually what we do when we create this class. I think you're right, it would be better to do it as a property

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt (hunk 1 above)
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt:130]: Updated to indicate that there were multiple deployments
